File: lib/java/openlinktoken/src/test/java/org/openlinktoken/attributes/person/SocialSecurityNumberAttributeTest.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.attributes.person;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SocialSecurityNumberAttributeTest {
    private SocialSecurityNumberAttribute ssnAttribute;

    @BeforeEach
    void setUp() {
        ssnAttribute = new SocialSecurityNumberAttribute();
    }

    @Test
    void getName_ShouldReturnSocialSecurityNumber() {
        assertEquals("SocialSecurityNumber", ssnAttribute.getName());
    }

    @Test
    void getAliases_ShouldReturnSocialSecurityNumberAliases() {
        String[] expectedAliases = { "SocialSecurityNumber", "NationalIdentificationNumber", "SSN" };
        assertArrayEquals(expectedAliases, ssnAttribute.getAliases());
    }

    @Test
    void normalize_ShouldFormatWithDashes() {
        assertEquals("123-45-6789", ssnAttribute.normalize("123456789"), "Should format without dashes");
        assertEquals("123-45-6789", ssnAttribute.normalize("123-45-6789"), "Should format with dashes");
        assertEquals("123-45-6789", ssnAttribute.normalize("123456789.0"), "Should format with decimal point");
        assertEquals("001-23-4567", ssnAttribute.normalize("1234567"), "Should format with leading zeros");
        assertEquals("001-23-4567", ssnAttribute.normalize("1234567.0"),
                "Should format with leading zeros and decimal point");
    }

    @Test
    void normalize_ShouldHandleEdgeCases() {
        assertEquals("1234567890", ssnAttribute.normalize("1234567890"), "Should return unchanged");
        assertEquals("12345678901.0", ssnAttribute.normalize("12345678901.0"),
                "Should drop decimal point even if exceeds length");
        assertEquals("12345678901", ssnAttribute.normalize("12345678901"),
                "Should return unchanged for long input without decimal");
        assertEquals("12345.0", ssnAttribute.normalize("12345.0"), "Should retain decimal point for short input");
        assertEquals("Unknown", ssnAttribute.normalize("Unknown"), "Should return non-numeric input unchanged");
        assertEquals("ABC-12-DEFG", ssnAttribute.normalize("ABC-12-DEFG"), "Should return non-numeric input unchanged");
    }

    @Test
    void normalize_ShouldHandleShortInputsWithoutCrashing() {
        assertEquals("123456", ssnAttribute.normalize("123456"), "Should handle 6-digit input without crashing");
        assertEquals("12345", ssnAttribute.normalize("12345"), "Should handle 5-digit input without crashing");
        assertEquals("1234", ssnAttribute.normalize("1234"), "Should handle 4-digit input without crashing");
        assertEquals("123", ssnAttribute.normalize("123"), "Should handle 3-digit input without crashing");
        assertEquals("12", ssnAttribute.normalize("12"), "Should handle 2-digit input without crashing");
        assertEquals("1", ssnAttribute.normalize("1"), "Should handle 1-digit input without crashing");
        assertEquals("", ssnAttribute.normalize(""), "Should handle empty input without crashing");
    }

    @Test
    void normalize_ShouldHandleSpaces() {
        assertEquals("123-45-6789", ssnAttribute.normalize("123 45 6789"), "Should normalize spaces to dashes");
        assertEquals("123-45-6789", ssnAttribute.normalize("123  45  6789"),
                "Should normalize multiple spaces to dashes");
        assertEquals("123-45-6789", ssnAttribute.normalize(" 123456789 "), "Should trim and format");
    }

    @Test
    void normalize_ShouldHandleMixedFormatting() {
        assertEquals("123-45-6789", ssnAttribute.normalize("123 45-6789"), "Should normalize mix of spaces and dashes");
        assertEquals("123-45-6789", ssnAttribute.normalize("123-45 6789"), "Should normalize mix of dashes and spaces");
        assertEquals("123-45-6789", ssnAttribute.normalize(" 123-456789"),
                "Should handle leading space and partial formatting");
    }

    @Test
    void normalize_ShouldHandleNullAndEmptyValues() {
        assertEquals(null, ssnAttribute.normalize(null), "Should return null for null input");
        assertEquals("", ssnAttribute.normalize(""), "Should return empty for empty input");
    }

    @Test
    void validate_ShouldReturnTrueForValidSSNs() {
        assertTrue(ssnAttribute.validate("223-45-6789"), "Valid SSN should be allowed");
        assertTrue(ssnAttribute.validate("223456789"), "Valid SSN without dashes should be allowed");
        assertTrue(ssnAttribute.validate("002-23-4567"), "Valid SSN with leading zeros should be allowed");
        assertTrue(ssnAttribute.validate("2234567"), "7-digit SSN should be allowed");
        assertTrue(ssnAttribute.validate("22345678"), "8-digit SSN should be allowed");
        assertTrue(ssnAttribute.validate("223456789.0"), "SSN with decimal should be allowed");
        assertTrue(ssnAttribute.validate("2234567.0"), "7-digit SSN with decimal should be allowed");
        assertTrue(ssnAttribute.validate("22345678.00"), "8-digit SSN with decimal should be allowed");
    }

    @Test
    void validate_ShouldReturnFalseForInvalidSSNs() {
        assertFalse(ssnAttribute.validate(null), "Null value should not be allowed");
        assertFalse(ssnAttribute.validate(""), "Empty value should not be allowed");
        assertFalse(ssnAttribute.validate("12345"), "Short SSN (5 digits) should not be allowed");
        assertFalse(ssnAttribute.validate("123456"), "Short SSN (6 digits) should not be allowed");
        assertFalse(ssnAttribute.validate("1234567890"), "Long SSN should not be allowed");
        assertFalse(ssnAttribute.validate("000-00-0000"), "Invalid sequence should not be allowed");
        assertFalse(ssnAttribute.validate("666-00-0000"), "SSN starting with 666 should not be allowed");
        assertFalse(ssnAttribute.validate("123-11-0000"), "All zeros in last group should not be allowed");
        assertFalse(ssnAttribute.validate("123-00-1234"), "All zeros in middle group should not be allowed");
        assertFalse(ssnAttribute.validate("000-45-6789"), "All zeros in first group should not be allowed");
        assertFalse(ssnAttribute.validate("900-45-6789"), "SSN starting with 900 should not be allowed");
        assertFalse(ssnAttribute.validate("999-45-6789"), "SSN starting with 999 should not be allowed");
        assertFalse(ssnAttribute.validate("ABCDEFGHI"), "Non-numeric should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-67AB"), "Mixed alphanumeric should not be allowed");
        assertFalse(ssnAttribute.validate("123.456.789"), "Multiple decimals should not be allowed");
        assertFalse(ssnAttribute.validate("123456789.123"),
                "Decimal with non-zero fractional part should not be allowed");
    }

    @Test
    void validate_ShouldReturnFalseForSpecificInvalidSSNs() {
        assertFalse(ssnAttribute.validate("001-01-0001"), "Invalid SSN 001-01-0001 should not be allowed");
        assertFalse(ssnAttribute.validate("001-01-0101"), "Invalid SSN 001-01-0101 should not be allowed");
        assertFalse(ssnAttribute.validate("001-01-9999"), "Invalid SSN 001-01-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("001-10-0110"), "Invalid SSN 001-10-0110 should not be allowed");
        assertFalse(ssnAttribute.validate("001-11-1111"), "Invalid SSN 001-11-1111 should not be allowed");
        assertFalse(ssnAttribute.validate("001-12-2334"), "Invalid SSN 001-12-2334 should not be allowed");
        assertFalse(ssnAttribute.validate("011-11-1111"), "Invalid SSN 011-11-1111 should not be allowed");
        assertFalse(ssnAttribute.validate("100-10-1000"), "Invalid SSN 100-10-1000 should not be allowed");
        assertFalse(ssnAttribute.validate("123-12-3123"), "Invalid SSN 123-12-3123 should not be allowed");
        assertFalse(ssnAttribute.validate("111-11-1112"), "Invalid SSN 111-11-1112 should not be allowed");
        assertFalse(ssnAttribute.validate("123-12-1234"), "Invalid SSN 123-12-1234 should not be allowed");
        assertFalse(ssnAttribute.validate("199-99-9999"), "Invalid SSN 199-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("899-99-9999"), "Invalid SSN 899-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("022-22-2222"), "Invalid SSN 022-22-2222 should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-7896"), "Invalid SSN 123-45-7896 should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-6788"), "Invalid SSN 123-45-6788 should not be allowed");
        assertFalse(ssnAttribute.validate("088-88-8888"), "Invalid SSN 088-88-8888 should not be allowed");
        assertFalse(ssnAttribute.validate("007-77-7777"), "Invalid SSN 007-77-7777 should not be allowed");
        assertFalse(ssnAttribute.validate("077-77-7777"), "Invalid SSN 077-77-7777 should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-6987"), "Invalid SSN 123-45-6987 should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-1234"), "Invalid SSN 123-45-1234 should not be allowed");
        assertFalse(ssnAttribute.validate("699-99-9999"), "Invalid SSN 699-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("111-11-1234"), "Invalid SSN 111-11-1234 should not be allowed");
        assertFalse(ssnAttribute.validate("222-33-4444"), "Invalid SSN 222-33-4444 should not be allowed");
        assertFalse(ssnAttribute.validate("111-22-1111"), "Invalid SSN 111-22-1111 should not be allowed");
        assertFalse(ssnAttribute.validate("111-22-2333"), "Invalid SSN 111-22-2333 should not be allowed");
        assertFalse(ssnAttribute.validate("111-11-9999"), "Invalid SSN 111-11-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("454-54-5454"), "Invalid SSN 454-54-5454 should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-6879"), "Invalid SSN 123-45-6879 should not be allowed");
        assertFalse(ssnAttribute.validate("055-55-5555"), "Invalid SSN 055-55-5555 should not be allowed");
        assertFalse(ssnAttribute.validate("123-45-6798"), "Invalid SSN 123-45-6798 should not be allowed");
        assertFalse(ssnAttribute.validate("299-99-9999"), "Invalid SSN 299-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("399-99-9999"), "Invalid SSN 399-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("499-99-9999"), "Invalid SSN 499-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("599-99-9999"), "Invalid SSN 599-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("799-99-9999"), "Invalid SSN 799-99-9999 should not be allowed");
        assertFalse(ssnAttribute.validate("111-22-3333"), "Invalid SSN 111-22-3333 should not be allowed");
        assertFalse(ssnAttribute.validate("111223333"), "Invalid SSN 111223333 (without dashes) should not be allowed");
    }

    @Test
    void normalize_ThreadSafety() throws InterruptedException {
        final int threadCount = 100;
        final CountDownLatch startLatch = new CountDownLatch(1);
        final CountDownLatch finishLatch = new CountDownLatch(threadCount);
        final CyclicBarrier barrier = new CyclicBarrier(threadCount);
        final List<String> results = Collections.synchronizedList(new ArrayList<>());
        final String testSSN = "123456789";

        for (int i = 0; i < threadCount; i++) {
            Thread thread = new Thread(() -> {
                try {
                    startLatch.await(); // Wait for all threads to be ready
                    barrier.await(); // Synchronize to increase contention

                    // Perform normalization
                    String result = ssnAttribute.normalize(testSSN);
                    results.add(result);

                    finishLatch.countDown();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
            thread.start();
        }

        startLatch.countDown(); // Start all threads
        finishLatch.await(15, TimeUnit.SECONDS); // Wait for all threads to complete

        // Verify all threads got the same result
        assertEquals(threadCount, results.size());
        for (String result : results) {
            assertEquals("123-45-6789", result);
        }
    }

    @Test
    void testSerialization() throws Exception {
        // Serialize the attribute
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(baos);
        oos.writeObject(ssnAttribute);
        oos.close();

        // Deserialize the attribute
        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
        ObjectInputStream ois = new ObjectInputStream(bais);
        SocialSecurityNumberAttribute deserializedAttribute = (SocialSecurityNumberAttribute) ois.readObject();
        ois.close();

        // Test various SSN values with both original and deserialized attributes
        String[] testValues = {
                "123456789",
                "123-45-6789",
                "001-23-4567",
                "001234567",
                "999-99-9999",
                "999999999"
        };

        for (String value : testValues) {
            assertEquals(
                    ssnAttribute.getName(),
                    deserializedAttribute.getName(),
                    "Attribute names should match");

            assertArrayEquals(
                    ssnAttribute.getAliases(),
                    deserializedAttribute.getAliases(),
                    "Attribute aliases should match");

            assertEquals(
                    ssnAttribute.normalize(value),
                    deserializedAttribute.normalize(value),
                    "Normalization should be identical for value: " + value);

            assertEquals(
                    ssnAttribute.validate(value),
                    deserializedAttribute.validate(value),
                    "Validation should be identical for value: " + value);
        }
    }
}

File: lib/python/openlinktoken/src/test/openlinktoken/attributes/person/social_security_number_attribute_test.py
# SPDX-License-Identifier: MIT

import pickle
from concurrent.futures import ThreadPoolExecutor, as_completed

import pytest

from openlinktoken.attributes.person.social_security_number_attribute import SocialSecurityNumberAttribute


class TestSocialSecurityNumberAttribute:
    """Test cases for SocialSecurityNumberAttribute class."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.ssn_attribute = SocialSecurityNumberAttribute()

    def test_get_name_should_return_social_security_number(self):
        """Test that get_name returns 'SocialSecurityNumber'."""
        assert self.ssn_attribute.get_name() == "SocialSecurityNumber"

    def test_get_aliases_should_return_social_security_number_aliases(self):
        """Test that get_aliases returns all supported social security number aliases."""
        expected_aliases = ["SocialSecurityNumber", "NationalIdentificationNumber", "SSN"]
        actual_aliases = self.ssn_attribute.get_aliases()
        assert actual_aliases == expected_aliases

    def test_normalize_should_format_with_dashes(self):
        """Test normalization formats SSNs with dashes."""
        assert self.ssn_attribute.normalize("123456789") == "123-45-6789"
        assert self.ssn_attribute.normalize("123-45-6789") == "123-45-6789"
        assert self.ssn_attribute.normalize("123456789.0") == "123-45-6789"
        assert self.ssn_attribute.normalize("1234567") == "001-23-4567"
        assert self.ssn_attribute.normalize("1234567.0") == "001-23-4567"

    def test_normalize_should_handle_edge_cases(self):
        """Test normalization handles edge cases."""
        assert self.ssn_attribute.normalize("1234567890") == "1234567890"
        assert self.ssn_attribute.normalize("12345678901.0") == "12345678901.0"
        assert self.ssn_attribute.normalize("12345678901") == "12345678901"
        assert self.ssn_attribute.normalize("12345.0") == "12345.0"
        assert self.ssn_attribute.normalize("Unknown") == "Unknown"
        assert self.ssn_attribute.normalize("ABC-12-DEFG") == "ABC-12-DEFG"

    def test_normalize_should_handle_short_inputs_without_crashing(self):
        """Test normalization handles short inputs without crashing."""
        assert self.ssn_attribute.normalize("123456") == "123456"
        assert self.ssn_attribute.normalize("12345") == "12345"
        assert self.ssn_attribute.normalize("1234") == "1234"
        assert self.ssn_attribute.normalize("123") == "123"
        assert self.ssn_attribute.normalize("12") == "12"
        assert self.ssn_attribute.normalize("1") == "1"
        assert self.ssn_attribute.normalize("") == ""

    def test_normalize_should_handle_spaces(self):
        """Test normalization handles spaces."""
        assert self.ssn_attribute.normalize("123 45 6789") == "123-45-6789"
        assert self.ssn_attribute.normalize("123  45  6789") == "123-45-6789"
        assert self.ssn_attribute.normalize(" 123456789 ") == "123-45-6789"

    def test_normalize_should_handle_mixed_formatting(self):
        """Test normalization handles mixed formatting."""
        assert self.ssn_attribute.normalize("123 45-6789") == "123-45-6789"
        assert self.ssn_attribute.normalize("123-45 6789") == "123-45-6789"
        assert self.ssn_attribute.normalize(" 123-456789") == "123-45-6789"

    def test_normalize_should_handle_null_and_empty_values(self):
        """Test normalization handles null and empty values."""
        assert self.ssn_attribute.normalize(None) is None
        assert self.ssn_attribute.normalize("") == ""

    def test_validate_should_return_true_for_valid_ssns(self):
        """Test validation returns true for valid SSNs."""
        assert self.ssn_attribute.validate("223-45-6789") is True
        assert self.ssn_attribute.validate("223456789") is True
        assert self.ssn_attribute.validate("2234567") is True
        assert self.ssn_attribute.validate("22345678") is True
        assert self.ssn_attribute.validate("223456789.0") is True
        assert self.ssn_attribute.validate("2234567.0") is True
        assert self.ssn_attribute.validate("22345678.00") is True

    def test_validate_should_return_false_for_invalid_ssns(self):
        """Test validation returns false for invalid SSNs."""
        assert self.ssn_attribute.validate(None) is False
        assert self.ssn_attribute.validate("") is False
        assert self.ssn_attribute.validate("12345") is False
        assert self.ssn_attribute.validate("123456") is False
        assert self.ssn_attribute.validate("1234567890") is False
        assert self.ssn_attribute.validate("000-00-0000") is False
        assert self.ssn_attribute.validate("666-00-0000") is False
        assert self.ssn_attribute.validate("123-11-0000") is False
        assert self.ssn_attribute.validate("123-00-1234") is False
        assert self.ssn_attribute.validate("000-45-6789") is False
        assert self.ssn_attribute.validate("900-45-6789") is False
        assert self.ssn_attribute.validate("999-45-6789") is False
        assert self.ssn_attribute.validate("ABCDEFGHI") is False
        assert self.ssn_attribute.validate("123-45-67AB") is False
        assert self.ssn_attribute.validate("123.456.789") is False
        assert self.ssn_attribute.validate("123456789.123") is False
        assert self.ssn_attribute.validate("001-01-0001") is False
        assert self.ssn_attribute.validate("001-01-0101") is False
        assert self.ssn_attribute.validate("001-01-9999") is False
        assert self.ssn_attribute.validate("001-10-0110") is False
        assert self.ssn_attribute.validate("001-11-1111") is False
        assert self.ssn_attribute.validate("001-12-2334") is False
        assert self.ssn_attribute.validate("011-11-1111") is False
        assert self.ssn_attribute.validate("100-10-1000") is False
        assert self.ssn_attribute.validate("123-12-3123") is False
        assert self.ssn_attribute.validate("111-11-1112") is False
        assert self.ssn_attribute.validate("123-12-1234") is False
        assert self.ssn_attribute.validate("199-99-9999") is False
        assert self.ssn_attribute.validate("899-99-9999") is False
        assert self.ssn_attribute.validate("022-22-2222") is False
        assert self.ssn_attribute.validate("123-45-7896") is False
        assert self.ssn_attribute.validate("123-45-6788") is False
        assert self.ssn_attribute.validate("088-88-8888") is False
        assert self.ssn_attribute.validate("007-77-7777") is False
        assert self.ssn_attribute.validate("077-77-7777") is False
        assert self.ssn_attribute.validate("123-45-6987") is False
        assert self.ssn_attribute.validate("123-45-1234") is False
        assert self.ssn_attribute.validate("699-99-9999") is False
        assert self.ssn_attribute.validate("111-11-1234") is False
        assert self.ssn_attribute.validate("222-33-4444") is False
        assert self.ssn_attribute.validate("111-22-1111") is False
        assert self.ssn_attribute.validate("111-22-2333") is False
        assert self.ssn_attribute.validate("111-11-9999") is False
        assert self.ssn_attribute.validate("454-54-5454") is False
        assert self.ssn_attribute.validate("123-45-6879") is False
        assert self.ssn_attribute.validate("055-55-5555") is False
        assert self.ssn_attribute.validate("123-45-6798") is False
        assert self.ssn_attribute.validate("299-99-9999") is False
        assert self.ssn_attribute.validate("399-99-9999") is False
        assert self.ssn_attribute.validate("499-99-9999") is False
        assert self.ssn_attribute.validate("599-99-9999") is False
        assert self.ssn_attribute.validate("799-99-9999") is False

    def test_normalize_thread_safety(self):
        """Test thread safety of normalize method."""
        thread_count = 100
        test_ssn = "123456789"
        expected_result = "123-45-6789"
        results = []

        def normalize_ssn():
            """Function to be executed by each thread."""
            try:
                result = self.ssn_attribute.normalize(test_ssn)
                return result
            except Exception as e:
                pytest.fail(f"Thread failed with exception: {e}")

        # Use ThreadPoolExecutor for better thread management
        with ThreadPoolExecutor(max_workers=thread_count) as executor:
            # Submit all tasks
            futures = [executor.submit(normalize_ssn) for _ in range(thread_count)]

            # Collect results
            for future in as_completed(futures, timeout=15):
                try:
                    result = future.result()
                    results.append(result)
                except Exception as e:
                    pytest.fail(f"Thread execution failed: {e}")

        # Verify all threads got the same result
        assert len(results) == thread_count, f"Expected {thread_count} results, got {len(results)}"

        for result in results:
            assert result == expected_result

    def test_serialization(self):
        """Test serialization and deserialization of SocialSecurityNumberAttribute."""
        # Serialize the attribute using pickle
        serialized_data = pickle.dumps(self.ssn_attribute)

        # Deserialize the attribute
        deserialized_attribute = pickle.loads(serialized_data)

        # Test various SSN values with both original and deserialized attributes
        test_values = ["123456789", "123-45-6789", "001-23-4567", "001234567", "999-99-9999", "999999999"]

        for value in test_values:
            # Test that attribute names match
            assert self.ssn_attribute.get_name() == deserialized_attribute.get_name()

            # Test that attribute aliases match
            assert self.ssn_attribute.get_aliases() == deserialized_attribute.get_aliases()

            # Test that normalization is identical
            assert self.ssn_attribute.normalize(value) == deserialized_attribute.normalize(value)

            # Test that validation is identical
            assert self.ssn_attribute.validate(value) == deserialized_attribute.validate(value)
